src/wallet/walletdb.cpp

```diff
@@ -1087,6 +1087,99 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         }
     }
 
+    // Upgrade to having a wallet hd key
+    // Find the keys which are used in single key internal and external descriptors with
+    // the pre-taproot output types
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS) &&
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) &&
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_GLOBAL_HD_KEY)
+        ) {
+        std::map<CExtPubKey, std::pair<std::map<std::string, int>, uint64_t>> descs_keys;
+        std::map<std::string, int> tmpl = {{"pkh(", 0}, {"sh(wpkh(", 0}, {"wpkh(", 0}};
+
+        // Find root xpubs used in pkh(), sh(wpkh()), and wpkh() descriptors
+        for (const auto& spkm : pwallet->GetAllScriptPubKeyMans()) {
+            const DescriptorScriptPubKeyMan* desc_spkm = dynamic_cast<DescriptorScriptPubKeyMan*>(spkm);
+            assert(desc_spkm);
+            LOCK(desc_spkm->cs_desc_man);
+            WalletDescriptor w_desc = desc_spkm->GetWalletDescriptor();
+
+            if (!w_desc.descriptor->IsRange()) {
+                continue;
+            }
+
+            std::set<CPubKey> desc_pubkeys;
+            std::set<CExtPubKey> desc_xpubs;
+            w_desc.descriptor->GetPubKeys(desc_pubkeys, desc_xpubs);
+            if (desc_xpubs.size() != 1) {
+                continue;
+            }
+            const CExtPubKey& xpub = *desc_xpubs.begin();
+
+            if (descs_keys.count(xpub) == 0) {
+                descs_keys.emplace(xpub, std::make_pair(tmpl, 0));
+            }
+
+            const std::string desc = w_desc.descriptor->ToString();
+            if (desc.find("pkh(") == 0) {
```

Comment:
> If all we need is to find the active HD master key, why not limit the search to `pkh(`? That has the nice benefit of there not potentially being manually inserted descriptors that could be ambiguous).

Users can still import `pkh(` descriptors. By checking all of the different types, we can be more sure about which key is active.

Reply:
> instead of this string parsing stuff, can we add some method to `Descriptor` to directly get what we need? Even an `enum` with descriptor types seems more clean.

I've changed it to use `OutputType` that apparently descriptors can report.